elastic/elastic-agent-shipper
# Refactor monitoring to make the shipper more beats-like, and compatible with Agent

## What does this PR do?

Part of https://github.com/elastic/elastic-agent-shipper/issues/267

This accomplishes a few things:

- changes the CLI and config layout to make the shipper's monitoring flags and config values interoperable with beats
- Completely rewrites the shipper monitoring to use the registry code used by other beats
- Adds a `/shipper` and `/stats` and `/state` endpoint to http monitoring
- Adds a config flag `http.expvar` that re-exports expvar metrics
- Fixes a bug where the shipper would not remove the gRPC endpoint on close
- fixes a bug where the shipper wasn't reporting a diagnostics callback for queue metrics

Note that this currently won't work with agent, as changes on the agent side are needed to make agent scrape metrics from the shipper.

## Why is it important?

Needed to integrate the shipper with other agent monitoring 

## Checklist



- [ ] I have added an entry in `CHANGELOG.md` or `CHANGELOG-developer.md`.

## Author's Checklist

- [ ] merge https://github.com/elastic/elastic-agent-system-metrics/pull/80

## How to test this PR locally

- pull and build
- Run the shipper in standalone mode with `http.enabled: true`
- cURL (or whatever) the `/`, `/shipper`, `/stats` and `/state` endpoints, make sure they're reporting valid data.
